crates/shielded_token/src/masp/shielded_wallet.rs

```diff
@@ -2032,11 +2286,184 @@ mod test_shielded_wallet {
                 };
 
                 wallet.add_note(
-                    create_note(asset_data.clone(), rewardless, pa),
+                    create_note(asset_data, rewardless, pa),
                     vk,
                 );
-                let rewards_est = wallet.estimate_next_epoch_rewards(&context, &vk).await.expect("Test failed");
-                assert_eq!(rewards_est, 2 * reward_rate * i128::from(principal));
+                let balance = wallet
+                    .compute_shielded_balance(&vk)
+                    .await
+                    .unwrap()
+                    .unwrap_or_else(ValueSum::zero);
+                let rewards_est = wallet.estimate_next_epoch_rewards(&context, &balance)
+                    .await.expect("Test failed");
+                assert_eq!(
+                    rewards_est,
+                    DenominatedAmount::native(
+                        Amount::from_masp_denominated_i128(
+                            (1 + i128::from(shield_at_previous_epoch)) * reward_rate * i128::from(principal),
+                            MaspDigitPos::Zero
+                        ).unwrap()
+                    ));
+            });
+        }
+
+        /// In this test, we have a single incentivized token, the native one,
+        /// shielded at MaspEpoch(2), owned by the shielded wallet.
+        /// The amount of owned token is the parameter `principal`.
+        ///
+        /// We add a conversion from MaspEpoch(1) to MaspEpoch(2)
+        /// which issues `reward_rate` nam tokens for our incentivized token.
+        ///
+        /// Optionally, the test also shields the same `principal` amount at
+        /// MaspEpoch(1).
```

Comment:
Again, maybe I'm confused about what's being tested, but shouldn't you shield at epoch 2 (the current epoch)?

Reply:
Yes but optionally I also want to shield an additional `principal` amount at the previous epoch to check that the rewards estimate logic holds